analysis/src/cratesio.rs

```diff
@@ -0,0 +1,108 @@
+
+use anyhow::{anyhow, Result};
+use crates_io_api::SyncClient;
+use guppy::graph::PackageMetadata;
+use tabled::Tabled;
+
+#[derive(Tabled)]
+pub struct CratesioReport {
+    pub name: String,
+    pub is_hosted: bool,
+    pub downloads: u64,
+    pub dependents: u64,
+}
+
+pub struct CratesioAnalyzer {
+    client: SyncClient,
+}
+
+impl CratesioAnalyzer {
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            client: SyncClient::new(
+                "User-Agent: Whackadep (https://github.com/diem/whackadep)",
+                std::time::Duration::from_millis(1000),
+            )?,
+        })
+    }
+
+    pub fn analyze_cratesio(self, package: &PackageMetadata) -> Result<CratesioReport> {
+        let name = package.name();
+        let is_hosted = package.source().is_crates_io();
+
+        if !is_hosted {
+            return Err(anyhow!("Crate {} not hosted on crates.io", name));
+        }
+
+        let crate_info = match self.client.get_crate(name) {
+            Ok(info) => info.crate_data,
+            Err(error) => return Err(error.into()),
+        };
+
+        // TODO: crates_io_api makes unnecessary request to page through all dependents,
+        // therefore, taking a long time
+        // when the total count is actually present in each result page
+        // make a PR to the upstream crate? or write custom http request by ourselves?
+        let dependents = match self.client.crate_reverse_dependencies(name) {
+            Ok(info) => info.meta.total,
+            Err(error) => return Err(error.into()),
+        };
+
+        println!("{} has {} dependents", name, dependents);
+
+        let cratesio_report = CratesioReport {
+            name: name.to_string(),
+            is_hosted,
+            downloads: crate_info.downloads,
+            dependents,
+        };
+
+        Ok(cratesio_report)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use guppy::MetadataCommand;
+    use std::path::PathBuf;
+
+    fn test_cratesio_analyzer() -> CratesioAnalyzer {
+        CratesioAnalyzer::new().unwrap()
+    }
+
+    #[test]
+    fn test_valid_package() {
+        let cratesio_analyzer = test_cratesio_analyzer();
+
+        let graph = MetadataCommand::new()
+            .current_dir(PathBuf::from("resources/test/valid_dep"))
+            .build_graph()
+            .unwrap();
+
+        let direct_dependencies: Vec<_> = graph
+            .query_workspace()
+            .resolve_with_fn(|_, link| {
+                // Collect direct dependencies of workspace packages.
+                let (from, to) = link.endpoints();
+                from.in_workspace() && !to.in_workspace()
+            })
+            .packages(guppy::graph::DependencyDirection::Forward)
+            .filter(|pkg| !pkg.in_workspace())
+            .collect();
+
+        // The test crate has only one dependency - libc
+        let libc = &direct_dependencies[0];
```

Comment:
Let's replace this with
```suggestion
        let libc = graph.packages().find(|p| p.name() == "libc").unwrap();
```

Reply:
done @bmwill